src/kv/KeyValueDB.h

```diff
@@ -226,77 +246,138 @@ class KeyValueDB {
   };
   typedef ceph::shared_ptr< WholeSpaceIteratorImpl > WholeSpaceIterator;
 
+  /// Iterator for column family which is created explicitly and not the default one.
+  class ColumnFamilyIteratorImpl {
+  public:
+    virtual int seek_to_first() = 0;
+    virtual int seek_to_last() = 0;
+    virtual int upper_bound(const std::string &after) = 0;
+    virtual int lower_bound(const std::string &to) = 0;
+    virtual bool valid() = 0;
+    virtual int next() = 0;
+    virtual int prev() = 0;
+    virtual std::string key() = 0;
+    virtual bufferlist value() = 0;
+    virtual bufferptr value_as_ptr() {
+      bufferlist bl = value();
+      if (bl.length()) {
+        return *bl.buffers().begin();
+      } else {
+        return bufferptr();
+      }
+    }
+    virtual int status() = 0;
+    virtual ~ColumnFamilyIteratorImpl() { }
+  };
+  typedef ceph::shared_ptr< ColumnFamilyIteratorImpl > ColumnFamilyIterator;
+
   class IteratorImpl : public GenericIteratorImpl {
-    const std::string prefix;
+    const std::string prefix;         //prefix or CF name
     WholeSpaceIterator generic_iter;
+    ColumnFamilyIterator cf_iter;
```

Comment:
It seems unnecessary to have this class wrap ColumnFamilyIterator; why not just make that class implement the GenericIteratorImpl interface and return it directly from get_iterator?

Reply:
Yes, ColumnFamilyIterator is not necessary. We will remove ColumnFamilyIteratorImpl, let RocksDBCFIteratorImpl inherit from GenericIteratorImpl.